### PR TITLE
Subsample state

### PR DESCRIPTION
**Pull request by mseeger**
*Issue #, if available:*

*Description of changes:*
Mechanism to subsample data for model-based multi-fidelity searchers. Switched on by default for MOBSTER, HyperTune, with max_size = 500 (in some experiments, this worked as well, maybe even a bit better, than larger limits or no limits).

Note that the same abstraction `state_converter` could also be used to drive multi-fidelity schedulers with single-fidelity searchers. The converter would just implement the logic to map to single fidelity (e.g., select last, average, ...). These converters can be chained as well.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: syne_tune/optimizer/schedulers/searchers/bayesopt/models/model_transformer.py [hunk 1]
```diff
@@ -136,6 +158,11 @@ class ModelStateTransformer:
     fitting only depends on ``state.trials_evaluations`` (observed data),
     not on other fields (e.g., pending evaluations).
 
+    If given, ``state_converter`` maps the state to another one which is then
+    passed to the model for fitting and predictions. One important use case is
+    filtering down data when model fitting is superlinear. Another is to convert
```
Comment (mseeger) [inline syne_tune/optimizer/schedulers/searchers/bayesopt/models/model_transformer.py:163]: The second use case is not yet implemented

File: syne_tune/optimizer/schedulers/searchers/bayesopt/tuning_algorithms/defaults.py [hunk 2]
```diff
@@ -25,3 +25,5 @@
 DEFAULT_NUM_INITIAL_CANDIDATES = 250
 
 DEFAULT_NUM_INITIAL_RANDOM_EVALUATIONS = 3
+
+DEFAULT_MAX_SIZE_DATA_FOR_MODEL = 500
```
Comment (mseeger) [inline syne_tune/optimizer/schedulers/searchers/bayesopt/tuning_algorithms/defaults.py:29]: According to experiments on NB201-CIFAR10 and NB301 with 8 workers and long max_wallclock_time, this works as well as larger values or no limit.

File: syne_tune/optimizer/schedulers/searchers/gp_fifo_searcher.py [hunk 3]
```diff
@@ -114,364 +78,6 @@ def create_initial_candidates_scorer(
         )
 
 
-def check_initial_candidates_scorer(initial_scoring: Optional[str]) -> str:
```
Comment (mseeger) [inline syne_tune/optimizer/schedulers/searchers/gp_fifo_searcher.py:117]: I split this file and created model_based_searcher.py.

File: syne_tune/optimizer/schedulers/searchers/model_based_searcher.py [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+        # here
+        if state_converter is not None:
+            state_converter.set_random_state(self.random_state)
+        if init_state is None:
+            init_state = TuningJobState.empty_state(self._hp_ranges_in_state())
+        self.state_transformer = ModelStateTransformer(
+            model_factory=model_factory,
+            init_state=init_state,
+            skip_optimization=skip_optimization,
+            state_converter=state_converter,
```
Comment (mseeger) [inline syne_tune/optimizer/schedulers/searchers/model_based_searcher.py:147]: New code. Apart, this is just copied from gp_fifo_searcher.py

**Review by aaronkl — DISMISSED** [review]
looks good

**Review by aaronkl — APPROVED** [review]

**Pull request merged**